File: src/day16.ts
function part1(data: string): string {
  const parts = data.split(',')
  let input = parts[0]
  const length = parseInt(parts[1] || '272')
  while (input.length < length) {
    input +=
      '0' +
      input
        .split('')
        .reverse()
        .join('')
        .replace(/0/g, 'z')
        .replace(/1/g, '0')
        .replace(/z/g, '1')
  }
  input = input.slice(0, length)

  let sum = ''
  while (sum.length % 2 == 0) {
    sum = ''
    for (let i = 0; i < input.length; i += 2) {
      sum += input[i] == input[i + 1] ? '1' : '0'
    }
    input = sum
  }
  return sum
}

function part2(data: string): string {
  let input = data
  const length = 35651584
  while (input.length < length) {
    input +=
      '0' +
      input
        .split('')
        .reverse()
        .join('')
        .replace(/0/g, 'z')
        .replace(/1/g, '0')
        .replace(/z/g, '1')
  }
  input = input.slice(0, length)

  let sum = ''
  while (sum.length % 2 == 0) {
    sum = ''
    for (let i = 0; i < input.length; i += 2) {
      sum += input[i] == input[i + 1] ? '1' : '0'
    }
    input = sum
  }
  return sum
}

export { part1, part2 }

File: 2017/day16.py
def part1(data, count=16):
    programs = [chr(i + ord("a")) for i in range(count)]
    run(programs, data)
    return "".join(programs)


def part2(data, count=16, iterations=1000000000):
    programs = [chr(i + ord("a")) for i in range(count)]
    done = set()
    i = 0
    while i < iterations:
        run(programs, data)
        if "".join(programs) in done:
            i = iterations - iterations % 36
            done.clear()
        done.add("".join(programs))
        i += 1
    return "".join(programs)


def run(programs, data):
    for move in data.split(","):
        if move[0] == "s":
            spin = int(move[1:])
            left, right = programs[-spin:], programs[:-spin]
            programs.clear()
            programs.extend(left + right)
        elif move[0] == "x":
            a, b = map(int, move[1:].split("/"))
            programs[a], programs[b] = programs[b], programs[a]
        elif move[0] == "p":
            a, b = [programs.index(x) for x in move[1:].split("/")]
            programs[a], programs[b] = programs[b], programs[a]
